aesd-char-driver/main.c

```diff
@@ -28,6 +30,92 @@ MODULE_LICENSE("Dual BSD/GPL");
 
 struct aesd_dev aesd_device;
 
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+    struct aesd_dev *dev = filp->private_data;
+    loff_t new_f_pos;
+    size_t total_size = 0;
+    size_t i;
+    size_t valid_entries = (dev->circ_buffer.full) ? AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED : dev->circ_buffer.in_offs;
+    
+    mutex_lock(&dev->lock);
+    /* Sum the sizes of all valid entries to compute total size */
+    for (i = 0; i < valid_entries; i++) {
+        size_t index = (dev->circ_buffer.out_offs + i) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        total_size += dev->circ_buffer.entry[index].size;
+    }
+    
+    switch (whence) {
+        case SEEK_SET:
+            new_f_pos = offset;
+            break;
+        case SEEK_CUR:
+            new_f_pos = filp->f_pos + offset;
+            break;
+        case SEEK_END:
+            new_f_pos = total_size + offset;
+            break;
+        default:
+            mutex_unlock(&dev->lock);
+            return -EINVAL;
+    }
+    
+    if (new_f_pos < 0) {
+        mutex_unlock(&dev->lock);
+        return -EINVAL;
+    }
+    
+    filp->f_pos = new_f_pos;
+    mutex_unlock(&dev->lock);
```

Comment:
@bharathvarmaCU18 as discussed in class, this could potentially be moved before the `switch` statement.